static/app/utils/useAsync.tsx

```diff
@@ -0,0 +1,382 @@
+import React, {useEffect, useState} from 'react';
+import * as Sentry from '@sentry/react';
+
+import {ResponseMeta} from 'sentry/api';
+import LoadingError from 'sentry/components/loadingError';
+import LoadingIndicator from 'sentry/components/loadingIndicator';
+import {t} from 'sentry/locale';
+import {metric} from 'sentry/utils/analytics';
+import getRouteStringFromRoutes from 'sentry/utils/getRouteStringFromRoutes';
+import useApi from 'sentry/utils/useApi';
+import useLocation from 'sentry/utils/useLocation';
+import useParams from 'sentry/utils/useParams';
+import useRoutes from 'sentry/utils/useRoutes';
+import PermissionDenied from 'sentry/views/permissionDenied';
+import RouteError from 'sentry/views/routeError';
+
+import RequestError from './requestError/requestError';
+
+type State = {
+  data: {[key: string]: any};
+  errors: {[key: string]: RequestError};
+  hasError: boolean;
+  isLoading: boolean;
+  isReloading: boolean;
+  remainingRequests: number;
+};
+
+type Result = {
+  renderComponent: (_child: React.ReactElement) => React.ReactElement;
+} & State;
+
+type Options = {
+  endpoints: Array<
+    [
+      string,
+      string,
+      {query?: string}?,
+      {allowError?: (_err) => void; disableEntireQuery?: boolean; paginate?: boolean}?
+    ]
```

Comment:
We should add names to these using the [named tuple feature](https://www.typescriptlang.org/play?ts=4.0.2#example/named-tuples)

```suggestion
  endpoints: Array<
    [
      key: string,
      url: string,
      urlOptions: {query?: string}?,
      requestOptions: {allowError?: (_err) => void; disableEntireQuery?: boolean; paginate?: boolean}?
    ]
```

Reply:
What would be really nice, is if we templateized the `useAsync` so you have to define the data types of each request, we could use the keys from that type to generate this array type as a fixed array that must include those data keys